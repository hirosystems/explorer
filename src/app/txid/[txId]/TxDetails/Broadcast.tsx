'use client';

import { FC } from 'react';
import * as React from 'react';
import { AiOutlineClockCircle } from 'react-icons/ai';

import { MempoolTransaction, Transaction } from '@stacks/stacks-blockchain-api-types';

import { KeyValueHorizontal } from '../../../../common/components/KeyValueHorizontal';
import { Value } from '../../../../common/components/Value';
import { toRelativeTime } from '../../../../common/utils/utils';
import { Box } from '../../../../ui/Box';
import { Flex } from '../../../../ui/Flex';
import { Icon } from '../../../../ui/Icon';
import { Tooltip } from '../../../../ui/Tooltip';
import { isInMempool } from '../utils';

export const Broadcast: FC<{
  tx: Transaction | MempoolTransaction;
}> = ({ tx }) => {
  if (!isInMempool(tx)) return null;

  const ts = tx.receipt_time;
  if (!ts) return null;

  const readableTs = `${new Date(ts * 1000).toLocaleTimeString()} ${new Date(
    ts * 1000
  ).toLocaleDateString()}`;

  return (
    <KeyValueHorizontal
      label={'Broadcast'}
      value={
        <>
          <Box>
            <Tooltip label={readableTs}>
              <Flex alignItems="center">
                <Icon as={AiOutlineClockCircle} size="16px" mr="4px" />
                <Value suppressHydrationWarning={true}>{toRelativeTime(ts * 1000)}</Value>
              </Flex>
            </Tooltip>
          </Box>
        </>
      }
      copyValue={readableTs}
    />
  );
};
