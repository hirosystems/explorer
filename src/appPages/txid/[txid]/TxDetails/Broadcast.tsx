import { AiOutlineClockCircle } from 'react-icons/ai';

import { MempoolTransaction, Transaction } from '@stacks/stacks-blockchain-api-types';
import { Box, Flex, Icon, Tooltip } from '@/ui/components';
import { toRelativeTime } from '@/common/utils';

import { KeyValueHorizontal } from '../../../common/components/KeyValueHorizontal';
import { Value } from '../../../common/components/Value';
import { isInMempool } from '../utils';

export function Broadcast({ tx }: { tx: Transaction | MempoolTransaction }) {
  if (!isInMempool(tx)) return null;

  const ts = tx.receipt_time;
  if (!ts) return null;

  const readableTs = `${new Date(ts * 1000).toLocaleTimeString()} ${new Date(
    ts * 1000
  ).toLocaleDateString()}`;

  return (
    <KeyValueHorizontal
      label="Broadcast"
      value={
        <Box>
          <Tooltip label={readableTs}>
            <Flex alignItems="center">
              <Icon as={AiOutlineClockCircle} size="16px" mr="4px" />
              <Value>{toRelativeTime(ts * 1000)}</Value>
            </Flex>
          </Tooltip>
        </Box>
      }
      copyValue={readableTs}
    />
  );
}
