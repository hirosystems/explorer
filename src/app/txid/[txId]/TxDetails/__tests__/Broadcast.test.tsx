import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';

import { Transaction } from '@stacks/stacks-blockchain-api-types';

import '../../../../../common/utils/test-utils/matchMedia.mock';
import { Broadcast } from '../Broadcast';

describe('<Broadcast />', () => {
  it('renders correctly', () => {
    const hardcodedDate = new Date('2023-01-01T12:00:00Z');

    const mockTx = {
      receipt_time: Math.floor(hardcodedDate.getTime() / 1000),
    } as unknown as Transaction;

    const { asFragment } = render(<Broadcast tx={mockTx} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
