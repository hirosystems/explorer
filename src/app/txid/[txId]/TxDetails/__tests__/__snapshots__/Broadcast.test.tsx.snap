// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Broadcast /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-12erdbu"
  >
    <div
      class="css-zbihu4"
    >
      Broadcast
    </div>
    <div
      class="css-1vv9c89"
    >
      <div
        class="css-0"
      >
        <div
          class="css-70qvj9"
        >
          <svg
            class="chakra-icon css-1a87sgk"
            fill="currentColor"
            focusable="false"
            height="4"
            viewBox="0 0 256 256"
            width="4"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm64-88a8,8,0,0,1-8,8H128a8,8,0,0,1-8-8V72a8,8,0,0,1,16,0v48h48A8,8,0,0,1,192,128Z"
            />
          </svg>
          <span
            class="chakra-text css-nvxyq8"
          >
            2 years ago
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
